### PR TITLE
fix(lsp): vim.lsp.start failed when an existing client has no workspace_folders

### PR DESCRIPTION
**Pull request by phanen**
## Problem
Seems a regression of https://github.com/neovim/neovim/pull/31340.

`nvim -l repro.lua`:
```lua
vim.lsp.start { cmd = { 'lua-language-server' }, name = 'lua_ls' }
vim.lsp.start { cmd = { 'lua-language-server' }, name = 'lua_ls', root_dir = 'foo' }

-- swaped case will be ok:
-- vim.lsp.start { cmd = { 'lua-language-server' }, name = 'lua_ls', root_dir = 'foo' }
-- vim.lsp.start { cmd = { 'lua-language-server' }, name = 'lua_ls' }
```

Failed with:
```
E5113: Error while calling lua chunk: /home/phan/b/neovim/runtime/lua/vim/lsp.lua:214: bad argument #1 to 'ipairs' (table expected, got nil)
stack traceback:
        [C]: in function 'ipairs'
        /home/phan/b/neovim/runtime/lua/vim/lsp.lua:214: in function 'reuse_client'
        /home/phan/b/neovim/runtime/lua/vim/lsp.lua:629: in function 'start'
        repro.lua:34: in main chunk
```

### CODE REVIEW

File: test/functional/plugin/lsp_spec.lua [hunk 1]
```diff
@@ -6121,6 +6121,23 @@ describe('LSP', function()
     end)
   end)
 
+  describe('vim.lsp.start', function()
+    it('can launch when an existing client has no workspace_folders', function()
+      exec_lua(create_server_definition)
+      exec_lua [[
```
Comment (lewis6991) [inline test/functional/plugin/lsp_spec.lua:6127]:
```suggestion
      exec_lua(function()
```

We can pass Lua functions to exec_lua now.

File: test/functional/plugin/lsp_spec.lua [hunk 2]
```diff
@@ -6121,6 +6121,23 @@ describe('LSP', function()
...
+          handlers = {
+            initialize = function(_, _, callback)
+              callback(nil, { capabilities = {} })
+            end,
+          },
+        })
+        vim.lsp.start { cmd = server.cmd, name = 'foo' }
+        vim.lsp.start { cmd = server.cmd, name = 'foo', root_dir = 'bar' }
+      end)
+    end)
```
Comment (justinmk) [inline test/functional/plugin/lsp_spec.lua:6138]: seems like this test is missing an assertion (such as `eq()`). does the test fail if you undo the change to `runtime/lua/vim/lsp.lua` ?

Comment (phanen) [reply test/functional/plugin/lsp_spec.lua:6137]: > seems like this test is missing an assertion (such as eq()). does the test fail if you undo the change to runtime/lua/vim/lsp.lua ?

Yes, it failed.

Maybe it's always a better practice to assert more thing than minimal reproduce...

File: test/functional/plugin/lsp_spec.lua [hunk 3]
```diff
@@ -6121,6 +6121,22 @@ describe('LSP', function()
     end)
   end)
 
+  describe('vim.lsp.start', function()
+    it('can launch when an existing client has no workspace_folders', function()
```
Comment (justinmk) [inline test/functional/plugin/lsp_spec.lua:6125]: The whole file is testing `vim.lsp.start` so this describe() is a bit unnecessary. Can just be a it() test.
```suggestion
  it('vim.lsp.start when existing client has no workspace_folders', function()
```

**Review by justinmk — APPROVED** [review]

File: test/functional/plugin/lsp_spec.lua [hunk 4]
```diff
@@ -6121,6 +6121,20 @@ describe('LSP', function()
     end)
   end)
 
+  it('vim.lsp.start when existing client has no workspace_folders', function()
```
[the remark this answers is not in the document]
Comment (phanen) [reply test/functional/plugin/lsp_spec.lua]: Thanks for guidance.

**Review by justinmk — APPROVED** [review]
